product_variant_default_code/models/product.py

```diff
@@ -98,44 +98,54 @@ class ProductTemplate(models.Model):
              '\nNote: make sure characters "[,]" do not appear in your '
              'attribute name')
 
+    def _get_attr_val_k(self, attr_value_id):
+        return attr_value_id.attribute_id.sequence
+
     def _get_default_mask(self):
+        if not self.attribute_line_ids:
+            return None
         attribute_names = []
         default_reference_separator = self.env[
             'ir.config_parameter'].get_param('default_reference_separator')
-        for line in self.attribute_line_ids:
+        # impossible to have an empty ir_config_parameter
+        if default_reference_separator == 'None':
+            default_reference_separator = ''
+        for line in sorted(self.attribute_line_ids, key=self._get_attr_val_k):
             attribute_names.append(u"[{}]".format(line.attribute_id.name))
         default_mask = ((self.code_prefix or '') +
                         default_reference_separator.join(attribute_names))
         return default_mask
 
+    def _is_automatic_mask(self):
+        return not self.user_has_groups(
+            'product_variant_default_code'
+            '.group_product_default_code')
+
     @api.model
     def create(self, vals):
         product = self.new(vals)
-        if (not vals.get('reference_mask') and product.attribute_line_ids or
-                not self.user_has_groups(
-                    'product_variant_default_code.group_product_default_code'
-                )):
+        if self._is_automatic_mask() or not vals.get('reference_mask'):
             vals['reference_mask'] = product._get_default_mask()
-        elif vals.get('reference_mask'):
+        else:
             sanitize_reference_mask(product, vals['reference_mask'])
         return super(ProductTemplate, self).create(vals)
 
+    def _mask_needs_update(self, vals):
+        return 'code_prefix' in vals or 'attribute_line_ids' in vals
+
     def write(self, vals):
-        product_obj = self.env['product.product']
-        if ('reference_mask' in vals and not vals['reference_mask'] or not
-                self.user_has_groups(
-                    'product_variant_default_code.group_product_default_code'
-                )):
-            if self.attribute_line_ids:
-                vals['reference_mask'] = self._get_default_mask()
         result = super(ProductTemplate, self).write(vals)
+        if (self._is_automatic_mask() and self._mask_needs_update(vals)) \
+                or ('reference_mask' in vals and not vals['reference_mask']):
+            for record in self:
+                # we write the new mask and the additional write
+                # will recompute the variant code
+                record.reference_mask = record._get_default_mask()
```

Comment:
I don't like too much that there are several nested writes. I prefer that you make this operation before, copying vals and adding the changed value, and call `super` record by record.

Reply:
Hi Pedro, I do not like a lot also the nested write, but in that case I think it's the simplest solution.
As for computing the "_get_default_mask" we need to have the attribute_line_ids wroten so the call to super must be done before getting the mask.

In the previous version if you modify the field "attribute_line_ids" the compute of the mask was done before the call to super and was generating a wrong mask with the previous value of the "attribute_line_ids".
What do you think?

Reply:
OK, thanks for the clarification. I didn't see that. Then the only solution is to nest the write, yes.

Reply:
Can you approve the review? thanks